tasks/src/main/scala/dagr/tasks/ScatterGather.scala

```diff
@@ -0,0 +1,175 @@
+   */
+  trait Scatterer[A] extends Task {
+    def scatters: Seq[A]
+  }
+
+  /** Provides a factory method to create a Scatter object from a Scatterer. */
+  object Scatter {
+    def apply[A](scatterer: Scatterer[A]): Scatter[A] = {
+      new ScattererWrapper(scatterer)
+    }
+  }
+
+
+  /**
+   * One of the two main types that implement Scatter/Gather, Scatter provides methods
+   * for mapping the currently scattered type to a new type by running a task in parallel
+   * on each of the scattered parts.
+   */
+  sealed trait Scatter[A] extends Task {
+    /**
+     * Produces a new Scatter from the existing scatter. Takes a function that maps an individual
+     * object of type A to a Task of type B.  During scatter/gather operation this function will
+     * be invoked with each A, to manufacture tasks of type B.
+     *
+     * The resulting Scatter[B] can be further mapped or gathered (or both!).
+     */
+    def map[B <: Task](f: A => B) : Scatter[B]
+
+    /**
+     * Produces a Gather object which will be used at runtime to manufacture a Task of type B
+     * which can gather things of type A.
+     */
+    def gather[B <: Task](f: Seq[A] => B) : Gather[A,B]
+  }
+
+
+  /**
+   * Represents a gather operation in a scatter/gather pipeline.
+   */
+  sealed trait Gather[A, B <: Task] extends Task {
+    private[ScatterGather] var sources: Option[Seq[A]] = None
+  }
+
+  /** Implementation of a Gather that performs a single All->One gather operation. */
+  private class SingleGather[A,B <: Task](f: Seq[A] => B) extends Gather[A,B] {
+    override def getTasks: Traversable[_ <: Task] = sources match {
+      case None    => throw new IllegalStateException("Gather.getTasks called before sources populated.")
+      case Some(a) => Some(f(a))
+    }
+  }
+
+
+  /**
+  * Implementation of a Scatter that is just a thinly veiled wrapper around the Scatterer
+  * being used to generate the set of scatters/partitions to operate on.
+  */
+  private class ScattererWrapper[A](scatterer: Scatterer[A]) extends Scatter[A] {
+    override def getTasks: Traversable[_ <: Task] = Some(scatterer)
+
+    override def gather[B <: Task](f: (Seq[A]) => B): Gather[A,B] =
+      throw new UnsupportedOperationException("gather not supported on an unmapped Scatter")
+
+    override def map[T <: Task](f: A => T): Scatter[T] = {
+      val sub = new PrimarySubScatter(scatterer, f)
+      this ==> sub
+      sub
+    }
+  }
+
+  /**
+  * Sub-trait of scatter that exists solely because we need two sub-types that share some common functionality.
+  * Provides the tracking of tasks generated by the Scatter, sub-Scatters that need to be invoked and wired in
+  * when the partitions are available and the set of Gathers to be performed on this stage of the Scatter.
+  */
+  private trait SubScatter[A, B <: Task] extends Scatter[B] {
+    val tasks   = new ListBuffer[B]
+    val subs    = new ListBuffer[SubScatter[B, _ <: Task]]
```

Comment:
Could this just be the following?
`val subs    = new ListBuffer[SecondarySubScatter[B, _ <: Task]]`
It may make it clearer that the `PrimarySubScatter` is the one that takes the initial scattered inputs while the `SecondarySubScatter` tasks simply chain along the "scatter" prior to a final gather.